bundles/org.eclipse.equinox.preferences/src/org/eclipse/core/internal/preferences/BundleStateScopeServiceFactory.java

```diff
@@ -0,0 +1,75 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Christoph Läubrich and others.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Christoph Läubrich - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.core.internal.preferences;
+
+import org.eclipse.core.internal.runtime.MetaDataKeeper;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.preferences.*;
+import org.eclipse.osgi.service.datalocation.Location;
+import org.osgi.framework.Bundle;
+import org.osgi.service.component.ComponentContext;
+import org.osgi.service.component.annotations.*;
+
+/**
+ * The Bundle State Scope supply bundles with two items:
+ * <ul>
+ * <li>{@link #getNode(String)} returns a node for the bundle workspace
+ * preferences</li>
+ * <li>{@link #getLocation()} returns the state location of the bundle that
+ * acquires the service</li>
+ * </ul>
+ * The service could be acquired for example with declarative services the
+ * following way:
+ *
+ * <pre>
+ * @Reference(target = "(type=bundle)")
+ * private IScopeContext scopeContext;
+ * </pre>
+ */
+@Component(service = IScopeContext.class, property = "type=bundle", scope = ServiceScope.BUNDLE)
+public class BundleStateScopeServiceFactory implements IScopeContext {
+
+	private Bundle usingBundle;
+	private IPath bundleLocation;
+
+	// this is not used, but makes sure the service is not available before a
+	// workspace url is set
+	@Reference(target = "(&(&(objectClass=org.eclipse.osgi.service.datalocation.Location)(type=osgi.instance.area))(url=*))")
+	Location workspaceLocation;
+
+	@Activate
+	void activate(ComponentContext context) {
+		usingBundle = context.getUsingBundle();
+	}
+
+	@Override
+	public String getName() {
+		return "bundle_state"; //$NON-NLS-1$
```

Comment:
It seems that the name here should match the type attribute on the service component.  I'm also not sure I like either name `bundle` or `bundle_state`.  Isn't this really the `instance.bundle` scope context?  Seems like we should relate it back to the `InstanceScope` API.  Perhaps even placing a constant in `InstanceScope` that describes the service property value and documents the published `IScopeContext` service.

It would seem that the `type` attribute name needs to be specified in a constant somewhere also, perhaps in `IScopeContext` stating that if the `IScopeContext` is registered as a service then it should specify a value for the `type` service property that describes the scope context.

Reply:
I never have had any use for the name (the abstract impl uses it as the name for the node so maybe we should simply return `InstanceScope.INSTANCE.getName/<bsn>` here?

Reply:
I have updated it with using better name/property constants.